src/osd/PrimaryLogPG.cc

```diff
@@ -3325,24 +3336,133 @@ struct SetManifestFinisher : public PrimaryLogPG::OpFinisher {
   }
 };
 
-void PrimaryLogPG::refcount_manifest(ObjectContextRef obc, object_locator_t oloc, hobject_t soid,
-                                     SnapContext snapc, bool get, Context *cb, uint64_t offset)
+bool PrimaryLogPG::check_has_ref_recent_snap(hobject_t soid, snapid_t snapid, chunk_info_t& ci)
+{
+  hobject_t clone_oid = soid;
+  clone_oid.snap = snapid;
+  bool has_reference = false;
+  ObjectContextRef cobc = get_object_context(clone_oid, false, NULL);
+  if (!cobc) {
+    dout(0) << __func__ << ": Can not find clone obc " << clone_oid << dendl;;
+    return false;
+  }
+  object_info_t& coi = cobc->obs.oi;
+  for (auto c : coi.manifest.chunk_map) {
+    if (!c.second.has_fingerprint()) {
+      // chunk
+      if (c.second.offset == ci.offset && c.second.length == ci.length &&
+	  c.second.oid.oid.name == ci.oid.oid.name) {
+	  has_reference = true;
+	  break;
+      } 
+    } else {
+      // dedup + chunk
+      if (c.second.oid.oid.name == ci.oid.oid.name) {
+	has_reference = true;
+	break;
+      }
+    }
+  }
+  dout(15) << __func__ << " target offset: " << ci.offset  << " oid: " << ci.oid 
+	  << " has references " << has_reference << dendl;
+  if (!has_reference) {
+    return false;
+  }
+  return true;
+}
+
+void PrimaryLogPG::dec_all_refcount_manifest(object_info_t& oi, OpContext* ctx)
```

Comment:
It seems like this overall process would be simpler and more reusable for the trim case if it worked in terms of manifest(s).  Like, if you have a ``set<hobject_t> chunks``, and the manifest object had a ``build_set()`` method, then you could do set operations with two manifests (e.g., current head manifest non-intersection with previous clone to get chunks to remove), or in the trim case, the to-trim set's non-intersection with the union of the previous and/or following manifests.

The current implementation seems to take two methods: one to check for an overlap, and then to check if the chunk appears anywhere.. but AFAICS the first part it redundant?